File: 프로그래머스/Level2/[3차] 방금그곡.js
// https://programmers.co.kr/learn/courses/30/lessons/17683

const replaceStr = (s) =>
	s
		.replace(/C#/g, 'c')
		.replace(/D#/g, 'd')
		.replace(/F#/g, 'f')
		.replace(/G#/g, 'g')
		.replace(/A#/g, 'a');

const getMin = (start, end) => {
	const [startH, startM] = start.split(':');
	const [endH, endM] = end.split(':');
	return (Number(endH) - Number(startH)) * 60 + Number(endM) - Number(startM);
};

const getStr = (melody, len) => {
	melody = replaceStr(melody);
	let result = '';
	const melodyLen = melody.length;
	const n = parseInt(len / melodyLen);
	const res = len % melodyLen;
	for (let i = 0; i < n; i++) {
		result += melody;
	}
	result += melody.slice(0, res);
	return result;
};

const solution = (m, musicinfos) => {
	const answer = [];
	m = replaceStr(m);
	for (let i = 0; i < musicinfos.length; i++) {
		const [start, end, name, melody] = musicinfos[i].split(',');
		const playMin = getMin(start, end);
		if (!getStr(melody, playMin).includes(m)) continue;
		answer.push([name, playMin]);
	}
	return answer.length === 0
		? '(None)'
		: answer.sort((a, b) => {
				return b[1] - a[1];
		  })[0][0];
};

File: 프로그래머스/Level2/[3차] 방금그곡.py
# https://programmers.co.kr/learn/courses/30/lessons/17683

def replaceStr(s):
    return s.replace('C#', 'c').replace('D#', 'd').replace('F#', 'f').replace('G#', 'g').replace('A#', 'a')

def getMin(start, end):
    startArr = start.split(':')
    endArr = end.split(':')
    return 60 * (int(endArr[0]) - int(startArr[0])) + int(endArr[1]) - int(startArr[1])

def getTotalMelody(melody, time):
    melodyLen = len(melody)
    n = time // melodyLen
    res = time % melodyLen
    return melody * n + melody[:res]

def solution(m, musicinfos):
    m = replaceStr(m)
    answer = []
    for info in musicinfos:
        arr = info.split(',')
        playMin = getMin(arr[0], arr[1])
        if m in getTotalMelody(replaceStr(arr[3]), playMin):
            answer.append((arr[2], playMin))
    if len(answer) == 0:
        return '(None)'
    answer.sort(key=lambda x:x[1], reverse=True)
    return answer[0][0]